File: 0819. Most Common Word/0819.java
// time complexity: O(N + B)
// space complexity: O(N + B)
// where N = length of paragraph (for split + scan) and B = total length of banned list.
class Solution {
    public String mostCommonWord(String paragraph, String[] banned) {
        // build a set of banned words for O(1) lookup
        Set<String> banSet = new HashSet<>(Arrays.asList(banned));
        // normalize to lowercase and split on non-letters
        String[] words = paragraph.toLowerCase().split("[^a-z]+");
        Map<String, Integer> count = new HashMap<>();
        String result = "";
        int maxFreq = 0;
        // count occurrences
        for (String w : words) {
            if (w.isEmpty() || banSet.contains(w)) continue;
            int freq = count.getOrDefault(w, 0) + 1;
            count.put(w, freq);
            // track the word with highest frequency
            if (freq > maxFreq) {
                maxFreq = freq;
                result = w;
            }
        }
        return result;
    }
}

File: 0819. Most Common Word/0819.py
# time complexity: O(N + B), where N is the length of the paragraph
# space compleixty: O(N + B), and B is the number of banned words
class Solution(object):
    def mostCommonWord(self, paragraph, banned):
        """
        :type paragraph: str
        :type banned: List[str]
        :rtype: str
        """
        # convert to lowercase
        para = paragraph.lower()
        # replace punctuation with spaces
        for ch in "!?',;.":
            para = para.replace(ch, " ")
        # split on whitespace to get words
        words = para.split()
        # build a set of banned words for fast lookup
        ban_set = set(banned)
        counts = {}
        # count each word if it is not banned
        for w in words:
            if w not in ban_set:
                counts[w] = counts.get(w, 0) + 1
        # return the word with the highest count
        # max(..., key=counts.get) finds the key with max value
        return max(counts, key=counts.get)